File: sampleCode/R/Endpoints/ImpactResultEndpoints.R
source(auth_env$Rest)
source(auth_env$Extensions)
source(auth_env$CommonFunctions)


setClass("ImpactResultEndpoints",
         slots = list(
           dummy = "character"
         ))

setClass("CsvReports",
         contains = "ImpactResultEndpoints")

# Create a generic function for Create
setGeneric("GetDetailedEconomicIndicators", function(object,impactRunId) standardGeneric("GetDetailedEconomicIndicators"))

# Define a method for the Create function with the CsvReports class
setMethod("GetDetailedEconomicIndicators", signature = "CsvReports", function(object,impactRunId){
  
  url <- paste0("api/v1/impact/results/ExportDetailEconomicIndicators/",impactRunId)
  
  csv <- ThrowIfNull(extension,GetResponseContent(rest, url = url, method_type = "GET"))
  
  return(csv)
  
  
})


# Create a generic function for Create
setGeneric("GetSummaryEconomicIndicators", function(object,impactRunId) standardGeneric("GetSummaryEconomicIndicators"))

# Define a method for the Create function with the CsvReports class
setMethod("GetSummaryEconomicIndicators", signature = "CsvReports", function(object,impactRunId){
  
  url <- paste0("api/v1/impact/results/SummaryEconomicIndicators/",impactRunId)
  
  csv <- ThrowIfNull(extension,GetResponseContent(rest, url = url, method_type = "GET"))
  
  return(csv)
  
})


# Create a generic function for Create
setGeneric("GetDetailedTaxes", function(object,impactRunId) standardGeneric("GetDetailedTaxes"))

# Define a method for the Create function with the CsvReports class
setMethod("GetDetailedTaxes", signature = "CsvReports", function(object,impactRunId){
  
  url <- paste0("api/v1/impact/results/DetailedTaxes/",impactRunId)
  
  csv <- ThrowIfNull(extension,GetResponseContent(rest, url = url, method_type = "GET"))
  
  return(csv)
  
})


# Create a generic function for Create
setGeneric("GetSummaryTaxes", function(object,impactRunId) standardGeneric("GetSummaryTaxes"))

# Define a method for the Create function with the CsvReports class
setMethod("GetSummaryTaxes", signature = "CsvReports", function(object,impactRunId){
  
  url <- paste0("api/v1/impact/results/SummaryTaxes/",impactRunId)
  
  csv <- ThrowIfNull(extension,GetResponseContent(rest, url = url, method_type = "GET"))
  
  return(csv)
  
})

setClass("EstimatedGrowthPercentageFilter",
         slots = list(
           DollarYear = "integer",
           Regions = "list",
           Impacts = "list",
           GroupNames = "list",
           EventNames = "list",
           EventTags = "list"
         ), 
         prototype = list(
           Regions = list(),
           Impacts = list(),
           GroupNames = list(),
           EventNames = list(),
           EventTags = list()
         ))

# Create a generic function for Create
setGeneric("GetEstimatedGrowthPercentage", function(object,impactRunId) standardGeneric("GetEstimatedGrowthPercentage"))

# Define a method for the Create function with the CsvReports class
setMethod("GetEstimatedGrowthPercentage", signature = "EstimatedGrowthPercentageFilter", function(object,impactRunId){

  url <- paste0("api/v1/impact/results/EstimatedGrowthPercentage/",impactRunId)
  
  print(url)
  
  growth_list_body <- convert_slots_into_list(object)
  
  # Convert JSON body to JSON format
  json_body_str <- toJSON(growth_list_body, auto_unbox = TRUE)
  
  print(json_body_str)
  
  response <- ThrowIfNull(extension,GetResponseContent(rest, url = url, method_type = "GET_BODY", response_body = json_body_str))
  
  return(response)

})

File: sampleCode/Python/endpoints/impact_result_endpoints.py
import requests
import logging

class ImpactType:
    Direct = 1
    Indirect = 2
    Induced = 3

class ImpactResultEndpoints:
    class CsvReports:
        @staticmethod
        def get_detailed_economic_indicators(impact_run_id, bearer_token):
            url = f"https://api.implan.com/api/v1/impact/results/ExportDetailEconomicIndicators/{impact_run_id}"
            headers = {"Authorization": f"Bearer {bearer_token}"}
            try:
                response = requests.get(url, headers=headers)
                response.raise_for_status()
                return response.text
            except requests.HTTPError as http_err:
                logging.error(f"HTTP error occurred: {http_err}")
                raise
            except Exception as err:
                logging.error(f"Other error occurred: {err}")
                raise

        @staticmethod
        def get_summary_economic_indicators(impact_run_id, bearer_token):
            url = f"https://api.implan.com/api/v1/impact/results/SummaryEconomicIndicators/{impact_run_id}"
            headers = {"Authorization": f"Bearer {bearer_token}"}
            try:
                response = requests.get(url, headers=headers)
                response.raise_for_status()
                return response.text
            except requests.HTTPError as http_err:
                logging.error(f"HTTP error occurred: {http_err}")
                raise
            except Exception as err:
                logging.error(f"Other error occurred: {err}")
                raise

        @staticmethod
        def get_detailed_taxes(impact_run_id, bearer_token):
            url = f"https://api.implan.com/api/v1/impact/results/DetailedTaxes/{impact_run_id}"
            headers = {"Authorization": f"Bearer {bearer_token}"}
            try:
                response = requests.get(url, headers=headers)
                response.raise_for_status()
                return response.text
            except requests.HTTPError as http_err:
                logging.error(f"HTTP error occurred: {http_err}")
                raise
            except Exception as err:
                logging.error(f"Other error occurred: {err}")
                raise

        @staticmethod
        def get_summary_taxes(impact_run_id, bearer_token):
            url = f"https://api.implan.com/api/v1/impact/results/SummaryTaxes/{impact_run_id}"
            headers = {"Authorization": f"Bearer {bearer_token}"}
            try:
                response = requests.get(url, headers=headers)
                response.raise_for_status()
                return response.text
            except requests.HTTPError as http_err:
                logging.error(f"HTTP error occurred: {http_err}")
                raise
            except Exception as err:
                logging.error(f"Other error occurred: {err}")
                raise

    class EstimatedGrowthPercentageFilter:
        def __init__(self, dollar_year, regions=None, impacts=None, group_names=None, event_names=None, event_tags=None):
            self.dollar_year = dollar_year
            self.regions = regions or []
            self.impacts = impacts or []
            self.group_names = group_names or []
            self.event_names = event_names or []
            self.event_tags = event_tags or []

        def to_dict(self):
            return {
                "DollarYear": self.dollar_year,
                "Regions": self.regions,
                "Impacts": [impact for impact in self.impacts],
                "GroupNames": self.group_names,
                "EventNames": self.event_names,
                "EventTags": self.event_tags,
            }

    @staticmethod
    def get_estimated_growth_percentage(impact_run_id, filter_, bearer_token):
        url = f"https://api.implan.com/api/v1/impact/results/EstimatedGrowthPercentage/{impact_run_id}"
        headers = {"Authorization": f"Bearer {bearer_token}", "Content-Type": "application/json"}
        payload = filter_.to_dict()
        logging.debug(f"Request URL: {url}")
        logging.debug(f"Request Headers: {headers}")
        logging.debug(f"Request Payload: {payload}")
        
        try:
            response = requests.post(url, json=payload, headers=headers)
            response.raise_for_status()
            
            logging.debug(f"Response Status Code: {response.status_code}")
            logging.debug(f"Response Headers: {response.headers}")
            logging.debug(f"Response Text: {response.text}")
            
            return response.json()
        except requests.exceptions.RequestException as e:
            logging.error(f"Request failed: {e}")
            return None